### DISCUSSION TITLE
System Configuration

### DISCUSSION THREAD

**Discussion by ls-1801** [nebulastream/nebulastream#44] [Ideas]
# System Configuration
Owner: @kisa7 @ls-1801 

## Background

System Configuration provides the infrastructure to load and distribute configuration parameters for system components.

## Goals

Each Sub-System has different needs in terms of configurable parameters. The System Configuration Library should provide common building blocks for other sub-systems to create reusable configurations without exposing module internals to other components.

In the future, we may decide to build applications with only a subset of the available system components, e.g., a Coordinator may or may not have a QueryCompiler or a QueryManager, allowing each application to pick and choose components requires the configuration to be as composable.

## Requirements

- component-specific configuration should reside in the components module
- configurable components expect a configuration upon construction
- The System Configuration Library provides common functionality: Configuration Types, Validation, Nesting of Configuration. Serialization: from yaml/json/command-line

## Open Points:

- [ ] Can all validation be handled by the System Configuration Library?
- [ ] Error handling during validation
- [ ] Implementation details: keep existing, more fancy c++, more

**Comment by kisa7** [comment]
I'd say we keep the existing configuration structure. 

I drew a diagram of the current hierarchy 
<img width="996" alt="Screen Shot 2024-07-09 at 13 01 53" src="https://github.com/nebulastream/nebulastream-public/assets/13611229/d81f1e23-eea8-4875-a6c1-4e9144c0622c">
The idea was to get rid of a TypedOption<T>, make BaseOption a template, and inherit ScalatOption<T> directly from the BaseOption<T>. However, the problem with this architecture is that BaseConfiguration class, which contains all the existing configurations, assumes each individual configuration to be abstract.

Points to consider:
1) Add a ClickHouse feature: Write "config-processed.yaml", once we process the documentation to simplify understanding for the user of what configs were used at the end. For example, if a user specifies 4 configuration options, the rest of the options will be assigned default values. It is nice for the user then to check the assigned values in the created file.
2) Runtime configuration
